File: quizzes/quizzesV4/quizzes9/01-problem.ts
export {};

type Users = {
    name: string;
    email: string;
};

type UsersByDomain = Record<string, Users['name'][]>;

const users: Users[] = [
    { name: 'Alice', email: 'alice@gmail.com' },
    { name: 'Bob', email: 'bob@yahoo.com' },
    { name: 'Carol', email: 'carol@gmail.com' },
];

const usersByDomain: UsersByDomain = {};

for (const user of users) {
    const domain: string = user.email.split('@')[1];
    (usersByDomain[domain] ??= []).push(user.name);
}

console.log(JSON.stringify({ usersByDomain }, null, 2));

File: bins1/01-problem.rb
# frozen_string_literal: true

require "json"

users = [
  { "name": "Alice", "email": "alice@gmail.com" },
  { "name": "Bob", "email": "bob@yahoo.com" },
  { "name": "Carol", "email": "carol@gmail.com" }
]

users_by_domain = users.each_with_object({}) do |user, hash|
  domain = user[:email].split('@').last
  (hash[domain] ||= []) << user[:name]
end

puts JSON.pretty_generate({ usersByDomain: users_by_domain })
